### PR TITLE
[agent-v1 03/18] feat(canvas): selection interaction mode + navigation primitives

### PR DESCRIPTION
**Pull request by CodeJuggernaut**
Merge after: #16145 (02b) - stacked on its branch; retargets to main as the stack lands.
Decisions: [Agent V1 Decision Log](https://app.notion.com/p/3c96d73d36508190b607d51ed9eeb3af)

## Summary

Slice 03 of the agent-v1 chain: the select-only picking mode + command guard, and the navigation primitives (fit-to-view, focus, viewport). Test-led per the chain rule; census vs merge-base: 36 files, +2039/-26.

## Changes

- **What**:
  - selectOnly gate (LGraphCanvas + the litegraphUtil guard doc + banner); focus/fit composables (useFocusNode, fitGraphToView, visibleCanvasViewport, DragAndScale.animateToBounds); command guards incl. history-layer undo/redo; both pin registrations; legacy Clear onclick; SelectionToolbox fail-closed.
  - Guard architecture (four layers): guarded command/composable call sites; the in-library pointer gates with gesture-time mode snapshots (a mode flip mid-gesture cannot invert an armed click or marquee); undo/redo at the history tracker (covers keyboard shortcuts, which bypass the command registry); node creation choked at addNodeOnGraph - plus four surface guards (document file drop, sidebar drop, workflow insert, createNode itself) and the node-replacement choke point (Errors-panel Replace Node/All).
  - Refusals are silent by design: while picking, creation/edit surfaces do nothing rather than raising errors. Suppressing the affordances themselves belongs to the mode-owner slice.

## Review Focus

- Guards reconstruct on current main (billing additions in useCoreCommands preserved - never replayed); vueNodesMode early-return precedes the gate (Nodes 2.0 unaffected); no agent ownership in this slice.
- Known unguarded, deferred to the mode-owner slice by design (boarded as blocking acceptance criteria there): the Vue-nodes pointer paths (click replace + drag), the widget-input conversion shims, and classic-canvas DOM widget inputs. Arming picking under Vue nodes requires guarding those first.
- The handleMeshFile refusal ordering (guard before upload) is read-verified; no suite reaches that private path (same carve-out as useCanvasDrop, whose guard is disclosed untested for lack of a suite).
- Notes:
  1. FENCE TRANSFER: the billing-zone fences (16067 bab0873d17 / 16100 74169f0bfb) and the 16001 import fence are vacuously satisfied at this cut - none is an ancestor of the stack base (merge-base a5a8ac872db7); their zones enter the tree at the retarget-onto-main merge, not in this diff.
  2. fitGraphToView lands consumer-less by ruling: wiring the FitView command to it is a flag-off behavior change; the child branch retarget brings its consumer.
  3. Attribution: FE-1756 = viewport/fit primitives, FE-1758 = selection mode + command guard (the tickets' live titles are the source of truth).
  4. SLICE-16 OBLIGATION (also boarded as a 16-lane acceptance item): workflow-lifecycle commands (new/open/load-default) are deliberately unguarded - a workflow swap EXITS picking; the mode's owner must implement that exit.

## Linear tickets

* [FE-1756](https://linear.app/comfyorg/issue/FE-1756) - canvas viewport geometry + fit-to-view primitives
* [FE-1757](https://linear.app/comfyorg/issue/FE-1757) - shared node navigation, focusNodeInstance + viewport-aware focusNode
* [FE-1758](https://linear.app/comfyorg/issue/FE-1758) - selectOnly interaction mode + ToggleSelectedNodes guard

### CODE REVIEW

**Comment by christian-byrne** [comment]
Automated fleet review (desktop operator, Christian-authorized)

- MERGE-AFTER-CHANGES.
- Fix gesture-mode snapshot bug.
- Add three regression tests.

<details><summary>Full context for agent readers</summary>

# PR #16186 — design-review packet (slice 03/18, canvas primitives)

**Subject**: [Comfy-Org/ComfyUI_frontend#16186](https://github.com/Comfy-Org/ComfyUI_frontend/pull/16186)
— "[agent-v1 03/18] feat(canvas): selection interaction mode + navigation primitives"
**Author**: CodeJuggernaut · **State**: OPEN, ready · **Head**: `ebe385c3ab60613bfc84a338d90443b5c7bee29d`
**Base**: `nathaniel/agent-v1/02b-crdt-write-leg` @ `d0f7be66d80966a61b6c9fc4c9c122f309842ffc`
**Size**: 36 files, +2039/−26
**Reviewed**: 2026-08-30 · report-only lane, no PR comments filed.

**Verdict: MERGE-AFTER-CHANGES** (2 cheap correctness items; the Vue-nodes limitation remains a
downstream arming condition).

---

## 1. What the PR actually does

This slice adds a default-off `selectOnly` canvas mode and guards classic-canvas editing while that
mode is active. It also adds viewport/focus helpers for later Agent navigation.

| Files | Kind | What changes |
|---|---|---|
| `src/lib/litegraph/src/LGraphCanvas.ts`, `DragAndScale.ts` | core canvas | select-only click/marquee behavior; viewport-aware animation |
| `src/composables/canvas/{fitGraphToView,useFocusNode,visibleCanvasViewport}.ts` | navigation | fit/focus/visible-viewport primitives |
| `src/composables/useCoreCommands.ts`, `usePaste.ts`; `src/scripts/changeTracker.ts` | command guards | block mutation, undo, and redo while picking |
| `src/scripts/app.ts`, `src/services/litegraphService.ts`, `src/utils/litegraphUtil.ts` | mutation choke points | block file/drop/node-creation paths |
| workflow, asset, drag, and node-replacement composables | shared callers | refuse their edits in select-only mode |
| 17 colocated `*.test.ts` files | tests | positive and negative guard, gesture, viewport, and focus coverage |

[nathaniel-pr-rationale: #16186] The PR says gesture-time mode snapshots prevent a mid-gesture
flag change from changing an armed click or marquee, and that selection is node-only. The
implementation does not fully preserve those two properties together: both marquee paths pass the
captured `pickingMode`, but eventually call `select(item)`, whose non-node filter reads the *live*
`this.selectOnly`. If picking starts, the mode flips off, and the gesture ends over a group, that
group can be selected.

The slice explicitly leaves Vue-nodes click, drag, link, resize, and widget-conversion paths
unguarded in `litegraphUtil.ts`. That disclosure is accurate. It is safe only while the later mode
owner refuses to arm selection mode under Vue nodes.

## 2. Alignment with our ADRs and CRDT recut

| Decision | Assessment |
|---|---|
| ADR-007 / ADR-010 | Aligned by non-contact: no op applier, follower, Y.Doc, or domain-store APPLY seam changes. |
| ADR-015 | Aligned: focus and selection remain presentation effects, not target-document authority. Later target navigation must still resolve an explicit workflow rather than fall back to the active canvas. |
| ADR-018 | No contact with the product-gated CRDT WebSocket surface. |

**CRDT-internals audit: clean.** The diff contains no import or changed path under
`extensions/agent/crdt/**` and no Yjs, Loro, state-vector, or `SCHEMA_VERSION` reference. Nothing
from the CRDT implementation that moved to the slices-13–17 recut is duplicated here.

## 3. Flag-off safety — CLEAN

The shared canvas code is reachable with `agent-in-app-experience=false`, but `selectOnly` defaults
false and the new branches are then behavior-preserving. There is no import-time listener,
registration, or visual change. This matches `reports/scorecards/flagoff-slice-03.md` and satisfies
BBC #149's user-facing-strict ruling without requiring Desktop discretion.

```text
Agent flag OFF ──▶ selectOnly=false ──▶ ordinary canvas behavior
Agent flag ON  ──▶ later owner arms mode ──▶ guarded node picking
```

## 4. Test coverage and CI

The slice has unusually broad unit coverage: click/marquee behavior, command and mutation choke
points, undo/redo, drop paths, fit/focus math, and positive outside-mode counterfactuals. At review
time GitHub reports **99 passing, 21 skipped, 1 cancelled** check; the cancelled check is
`cursor-review / edge-case`, not an executed test failure.

The missing regression is the defect above: no test starts a click/classic marquee/live marquee
over a group in picking mode, flips the mode before release, and verifies the group remains
unselected. Browser-observable selection-mode coverage also remains for the mounting slice.

## 5. Verdict

### MERGE-AFTER-CHANGES

1. **Make the node-only filter use the captured gesture mode.** Carry `pickingMode` into the final
   selection operation instead of allowing `select()` to consult the live `selectOnly` value.
2. **Add the discriminating group regressions.** Cover click, classic marquee, and live marquee
   with picking true at pointer-down, false at pointer-up; groups must not land in selection.

**Downstream gate, not a third change in this slice:** do not arm picking under Vue nodes until the
explicitly listed Vue-node mutation paths are guarded. [nathaniel-pr-rationale: #16186] The PR
assigns that work to the mode-owner slice; the slices-13–17 plan assigns final navigation and
activation integration later in the chain.

## Sources

Direct PR diff and checks at the exact head; `plans/wave-plan-2026-08-28.md`;
`reports/audit/feh4-slice-plan-audit.md`; ADR-007, ADR-010, ADR-015, ADR-018; BBC #149 ruling;
`reports/scorecards/flagoff-slice-03.md`. Nathaniel's linked CRDT design material was not opened.

## Glossary

- **Select-only / picking mode**: temporary canvas mode in which users may choose nodes but may not edit the graph.
- **Gesture snapshot**: the mode value captured at pointer-down and retained until pointer-up.
- **Vue nodes**: the newer Vue-rendered canvas-node path, distinct from classic LiteGraph pointer handling.
- **CRDT**: conflict-free replicated data type; the shared workflow-state mechanism owned by the recut slices.
- **BBC #149**: Christian's ruling that flag-off strictness applies to user-facing regressions, with Desktop discretion for cosmetic shared changes.


</details>

File: src/lib/litegraph/src/LGraphCanvas.ts [hunk 1]
```diff
@@ -4600,6 +4674,7 @@ export class LGraphCanvas implements CustomEventDispatcher<LGraphCanvasEventMap>
   select<TPositionable extends Positionable = LGraphNode>(
     item: TPositionable
   ): void {
+    if (this.selectOnly && !(item instanceof LGraphNode)) return
```
[the remark this answers is not in the document]
Comment (christian-byrne) [reply src/lib/litegraph/src/LGraphCanvas.ts]: Already fixed: gesture snapshots govern marquee policy and item filtering.

<details><summary>Full context for agent readers</summary>

`_handleMultiSelect` and `handleLiveSelect` pass captured `pickingMode` into `select(item, pickingMode)`. `processSelect` also rejects non-node items from captured `picking`. Turning off the live flag cannot admit groups or reroutes from an armed picking gesture.

</details>

**Comment by christian-byrne** [comment]
Gate 6 FE-1756 blocker fixed.

- Generic viewport inset API; no agent dependency.
- Empty, nonzero, and clamped cases green.

<details><summary>Full context for agent readers</summary>

Commit `4382e821394aeeb62e2eaee85a183c5ca5506490` adds an optional four-edge `CanvasViewportInset` input to `visibleCanvasViewport()`. Insets apply after the existing graph-panel intersection, negative values clamp to zero, and oversized values collapse the viewport within canvas bounds. Existing callers remain unchanged; slice 03 stays flag-gated and panel-agnostic.

Evidence: 10 affected unit tests passed across `visibleCanvasViewport.test.ts` and `useFocusNode.test.ts`; `pnpm typecheck` passed; targeted ESLint and oxfmt checks passed. Full `pnpm lint` exited 0 with pre-existing warnings only.

</details>

**Comment by christian-byrne** [comment]
Superseded — closing per Christian's direction.

- v2 merged to main in #16198 (f954e479); this slice's surface now lives on main.
- Salvage flows through new main-based PRs (e.g. #16340), not this stack.
- Reopen if unique content surfaces that main lacks.

<details><summary>Full context for agent readers</summary>

Program pivot 2026-08-31: the 17-slice stack targeting the v1 integration branch is superseded by the #16198 v2 merge to main @f954e479. Per program/reorientation-plan.md, all remaining value (tests, QA artifacts, invariant checks, refactors) ports to main-based successor PRs. Pre-close dossier cross-check completed for this PR; unique unsalvaged content (if any) was filed as a salvage row in the program todo before close. This closed PR remains a readable source; review/QA deliverables are posted only on open successor PRs.

</details>

**Comment by christian-byrne** [comment]
This slice is reference-only. Accepted work is routed to [#16198](https://github.com/Comfy-Org/ComfyUI_frontend/pull/16198) and [#16401](https://github.com/Comfy-Org/ComfyUI_frontend/pull/16401).

<details><summary>Full context for agent readers</summary>

Per Christian's main-target close ruling, [#16186](https://github.com/Comfy-Org/ComfyUI_frontend/pull/16186) (03 — canvas selection and navigation) is no longer a landing vehicle. It remains readable implementation and review provenance. Broad selection-mode blocking was deliberately rejected. If a genuinely unaccounted delta is found, reopen this source or file a bounded main-based carrier rather than reviving the stacked merge path.

[Decision record](https://github.com/christian-byrne/in-app-agent-program/blob/main/decisions/2026-09-02-slice-prs-reference-only.md).

</details>
